switchboard/manager.py

```diff
@@ -214,9 +214,20 @@ def as_request(self, user=None, ip_address=None):
         return MockRequest(user, ip_address)
 
     def version_switch(self, switch):
-        if hasattr(switch, 'save_version'):
-            user = self.context.get('user')
-            switch.save_version(username=user.username if user else '')
+        # Try to get the username from both User objects and user dicts.
+        try:
+            user = self.context.get('user', {})
+            if hasattr(user, 'username'):
+                username = user.username
+            else:
+                username = user.get('username', '')
+        except:
```

Comment:
That's probably true for the code below.  In this block about the only thing that could go wrong is self.context not existing or being None such that the .get() fails. Everything else that could go wrong is defensively handled.

I agree that being failsafe is important here but bare excepts are creepy.